src/task/FinishGatheringICECandidatesTask.ts

```diff
@@ -62,98 +62,103 @@ export default class FinishGatheringICECandidatesTask extends BaseTask {
+            }`
+          );
+          // Ice candidate arrives, do not need to wait anymore.
+          // https://webrtcglossary.com/trickle-ice/
+          if (event.candidate) {
+            if (DefaultSDP.isRTPCandidate(event.candidate.candidate)) {
+              this.context.iceCandidates.push(event.candidate);
+            }
+
+            // Could there be a case the candidate is not written to SDP ?
+            if (this.context.turnCredentials && this.context.iceCandidates.length >= 1) {
+              this.context.logger.info('gathered at least one relay candidate');
+              this.removeEventListener();
+              resolve();
+              return;
+            }
           }
-        }
 
-        // Ice candidate gathering is complete, additional barrier to make sure sdp contain an ice candidate.
-        // TODO: Could there be a race where iceGatheringState is flipped after this task is run ? This could only be handled if ice state is monitored persistently.
-        if (this.context.peer.iceGatheringState === 'complete') {
-          this.context.logger.info('done gathering ice candidates');
-          this.removeEventListener();
-          if (
-            !new DefaultSDP(this.context.peer.localDescription.sdp).hasCandidates() ||
-            this.context.iceCandidates.length === 0
-          ) {
-            reject(new Error('no ice candidates were gathered'));
-          } else {
-            resolve();
+          // Ice candidate gathering is complete, additional barrier to make sure sdp contain an ice candidate.
+          // TODO: Could there be a race where iceGatheringState is flipped after this task is run ? This could only be handled if ice state is monitored persistently.
+          if (this.context.peer.iceGatheringState === 'complete') {
+            this.context.logger.info('done gathering ice candidates');
+            this.removeEventListener();
+            if (
+              !new DefaultSDP(this.context.peer.localDescription.sdp).hasCandidates() ||
+              this.context.iceCandidates.length === 0
+            ) {
+              reject(new Error('no ice candidates were gathered'));
+            } else {
+              resolve();
+            }
           }
-        }
-      };
+        };
 
-      // SDK does not catch candidate itself and send to sever. Rather, WebRTC handles candidate events and writes candidate to SDP.
-      this.context.peer.addEventListener('icecandidate', this.context.iceCandidateHandler);
-      this.startTimestampMs = Date.now();
-    });
+        // SDK does not catch candidate itself and send to sever. Rather, WebRTC handles candidate events and writes candidate to SDP.
+        this.context.peer.addEventListener('icecandidate', this.context.iceCandidateHandler);
+        this.startTimestampMs = Date.now();
+      });
+    } catch (error) {
+      throw error;
+    } finally {
+      this.context.iceGatheringDurationMs = Math.round(Date.now() - this.startTimestampMs);
```

Comment:
If the task returns at line 71 or 92 or throws anywhere before `this.startTimestampMs` is set, the `iceGatheringDurationMs` value might be `NaN`. You can solve in this way:

1. Use the try/catch statement to wrap only the promise.
2. To be defensive, also check if the `startTimestampMs` exists.

```
try {
  await new Promise<void>((resolve, reject) => {
     ...
  });
} catch (error) {
  throw error;
} finally {
  if (this.startTimestampMs) {
     this.context.iceGatheringDurationMs  = Math.round(Date.now() - this.startTimestampMs);
  }
}
```